bin/cron/retention_stats

```diff
@@ -9,45 +9,81 @@ require_relative '../../dashboard/config/environment'
   drop_off = UserScript.
     where(last_progress_at: (Time.now - 4 * ONE_WEEK)..(Time.now - 2 * ONE_WEEK))
+  drop_off_with_levels = drop_off.
+    joins('INNER JOIN user_levels ON user_levels.user_id = user_scripts.user_id AND user_levels.script_id = user_scripts.script_id')
+  drop_off_with_levels_and_stages = drop_off_with_levels.
+    joins('INNER JOIN script_levels ON script_levels.script_id = user_levels.script_id AND script_levels.level_id = user_levels.level_id')
+
+  # Hashes mapping script_id to the number of users starting / finishing the script.
   script_starts = drop_off.group(:script_id).count
   script_finishes = drop_off.where.not(completed_at: nil).group(:script_id).count
 
-  drop_off_with_levels = drop_off.
-    joins("INNER JOIN user_levels ON user_levels.user_id = user_scripts.user_id AND user_levels.script_id = user_scripts.script_id")
-  # Determine, for each (user, script), the number of levels the user attempted. Then construct a
-  # hash of hashes of the form {script_id: {level_count: number_of_users}}.
-  levels = Hash.new {|hash, key| hash[key] = Hash.new(0)}
+  # Determine, for each (user, script), the number of levels attempted. Then increment the
+  # appropriate index in the script_level_counts hash.
   drop_off_with_levels.
     group('user_levels.user_id', 'user_levels.script_id').
     count.
-    each{|pair, count| levels[pair.second][count] += 1}
-  # Determine, for each (user, script, stage), the number of levels attempted. Then construct a hash
-  # of hashes of hashes of the form {script_id: {stage_id: {level_count: number_of_users}}}.
-  stages = Hash.new {|hash, key| hash[key] = Hash.new {|hash, key| hash[key] = Hash.new {|hash, key| hash[key] = 0}}}
-  drop_off_with_levels.
-    joins("INNER JOIN script_levels ON script_levels.script_id = user_levels.script_id AND script_levels.level_id = user_levels.level_id").
+    each_pair do |user_id_script_id, level_count|
+      script_level_counts[user_id_script_id.second][level_count] += 1
+    end
+  # Determine, for each (user, script), the number of stages attempted. Then increment the
+  # appropriate index in the script_stage_counts hash.
+  drop_off_with_levels_and_stages.
+    group('user_levels.user_id', 'script_levels.script_id', 'script_levels.stage_id').
+    distinct.    
+    count('script_levels.stage_id').
+    each_pair do |user_id_script_id, stage_count|
+      script_stage_counts[user_id_script_id.second][stage_count] += 1
+    end
+  # Determine, for each (user, script, stage), the number of levels attempted. Then increment the
+  # appropriate index in the stage_level_counts hash.
+  drop_off_with_levels_and_stages.
     group('user_levels.user_id', 'script_levels.script_id', 'script_levels.stage_id').
     count.
-    each{|triple, count| stages[triple.second][triple.third][count] += 1}
+    each_pair do |user_id_script_id_stage_id, level_count|
+      stage_level_counts[user_id_script_id_stage_id[2]][level_count] += 1
```

Comment:
nit: for parallelism, it would be nice to use either `.third` here, or `user_id_script_id[1]` above